### PR TITLE
Add multiclass_nms3 GPU kernel

### PR DESCRIPTION
**Pull request by Tom-Zheng**
### PR types

New features
### PR changes

OPs
### Description

In this PR, we add a GPU kernel for multiclass_nms3 op, which could greatly speed up model evaluation for detection models.

We benchmarked in PP-YOLOE+ evaluation, with `ppyoloe_plus_crn_l_80e_coco.yml` config. 
Setting: A100-PCIE-80GB; batch_size=32; evaluate size = 640 x 640.
Problem size of NMS OP: shape of bbox: [32, 8400, 4]; shape of scores: [32, 80, 8400]
Other parameters:
- 'nms_top_k': 1000,
- 'keep_top_k': 300,
- 'score_threshold': 0.01,
- 'nms_threshold': 0.7

Benchmark result:
NMS OP time: 2295 ms (CPU) -> 0.267 ms (GPU) ; speedup: 8595.5x

### CODE REVIEW

**Review by lyuwenyu — DISMISSED** [review]

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu [hunk 1]
```diff
@@ -0,0 +1,1141 @@
...
+      nmsedBoxes[i * 4] = clipBoxes ? saturate(xMin) : xMin;
+      // clipped bbox ymin
+      nmsedBoxes[i * 4 + 1] = clipBoxes ? saturate(yMin) : yMin;
+      // clipped bbox xmax
+      nmsedBoxes[i * 4 + 2] = clipBoxes ? saturate(xMax) : xMax;
+      // clipped bbox ymax
+      nmsedBoxes[i * 4 + 3] = clipBoxes ? saturate(yMax) : yMax;
+      nmsedIndices[i] = bboxId >> 2;
+      nmsedValidMask[i] = 1;
+      atomicAdd(&numDetections[i / keepTopK], 1);
```
Comment (shaojiewang) [inline paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu]: TODO: we may also need a deterministic version.

**Review by shaojiewang — COMMENTED** [review]
Could you demonstrate some test cases with perf number?

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu (hunk 1 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu]: The `atomicAdd` is performed on integer, so there is no determinism issue.

Comment (Tom-Zheng) [reply paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu]: We benchmarked in PP-YOLOE+ evaluation, with `ppyoloe_plus_crn_l_80e_coco.yml` config. 
Setting: A100-PCIE-80GB; batch_size=32; evaluate size = 640 x 640.
Problem size of NMS OP: shape of bbox: [32, 8400, 4]; shape of scores: [32, 80, 8400]
Other parameters:
- 'nms_top_k': 1000,
- 'keep_top_k': 300,
- 'score_threshold': 0.01,
- 'nms_threshold': 0.7

Benchmark result:
NMS OP time: 2295 ms (CPU) -> 0.267 ms (GPU) ; speedup: 8595.5x

Comment (shaojiewang) [reply paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu]: > 



> The `atomicAdd` is performed on integer, so there is no determinism issue.

sure, got it.

Comment (shaojiewang) [reply paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu]: > We benchmarked in PP-YOLOE+ evaluation, with `ppyoloe_plus_crn_l_80e_coco.yml` config. Setting: A100-PCIE-80GB; batch_size=32; evaluate size = 640 x 640. Problem size of NMS OP: shape of bbox: [32, 8400, 4]; shape of scores: [32, 80, 8400] Other parameters:
> 
> * 'nms_top_k': 1000,
> * 'keep_top_k': 300,
> * 'score_threshold': 0.01,
> * 'nms_threshold': 0.7
> 
> Benchmark result: NMS OP time: 2295 ms (CPU) -> 0.267 ms (GPU) ; speedup: 8595.5x

could you put it into PR description?

**Review by shaojiewang — DISMISSED** [review]
lgtm

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu (hunk 1 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu]: Done.

File: python/paddle/fluid/tests/unittests/test_multiclass_nms_op.py [hunk 2]
```diff
@@ -797,7 +797,8 @@ def setUp(self):
         }
 
     def test_check_output(self):
-        self.check_output()
+        place = paddle.CPUPlace()
```
Comment (Xreki) [inline python/paddle/fluid/tests/unittests/test_multiclass_nms_op.py:807]: 这里为啥改成只检查CPU呢

File: python/paddle/fluid/tests/unittests/test_multiclass_nms_op.py [hunk 3]
```diff
@@ -807,6 +808,90 @@ def set_argument(self):
         self.score_threshold = 2.0
 
 
+@unittest.skipIf(
+    not core.is_compiled_with_cuda(), "core is not compiled with CUDA"
+)
+class TestMulticlassNMS3OpGPU(TestMulticlassNMS3Op):
```
Comment (Xreki) [inline python/paddle/fluid/tests/unittests/test_multiclass_nms_op.py]: 为何要另外构造测试方式，原单测框架不能检查吗？

File: paddle/fluid/operators/detection/multiclass_nms_op.cc [hunk 4]
```diff
@@ -614,6 +614,13 @@ class MultiClassNMS3Op : public MultiClassNMS2Op {
                    const framework::VariableNameMap& outputs,
                    const framework::AttributeMap& attrs)
       : MultiClassNMS2Op(type, inputs, outputs, attrs) {}
+
+ protected:
+  phi::KernelKey GetExpectedKernelType(
```
Comment (Xreki) [inline paddle/fluid/operators/detection/multiclass_nms_op.cc:619]: phi体系下，指定Kernel选择的数据类型方式，可参考https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/phi/api/yaml/legacy_ops.yaml#L129

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu [hunk 5]
```diff
@@ -0,0 +1,1141 @@
...
+
+#ifndef PADDLE_WITH_HIP
+
+#include "paddle/phi/kernels/multiclass_nms3_kernel.h"
+#include "paddle/phi/kernels/impl/multiclass_nms3_kernel_impl.h"
+
+#include <cub/cub.cuh>
+#include "cuda.h"  // NOLINT
+
+#include "paddle/fluid/framework/tensor_util.h"
```
Comment (Xreki) [inline paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu]: 不要include fluid下面的头文件

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu [hunk 6]
```diff
@@ -0,0 +1,1141 @@
...
+      reinterpret_cast<ValueT*>(NULL),
+      num_items,     // # items
+      num_segments,  // # segments
+      reinterpret_cast<int*>(NULL),
+      reinterpret_cast<int*>(NULL));
+  return temp_storage_bytes;
+}
+
+template <typename T>
+size_t detectionForwardBBoxDataSize(int N, int C1) {
```
Comment (Xreki) [inline paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu]: 函数命名请采用大写字母的驼峰式，且最好是动词词组，`detectionForwardBBoxDataSize` -> `CalcDetectionForwardBBoxDataSize`，下同

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu [hunk 7]
```diff
@@ -0,0 +1,1141 @@
...
+        afterNMS_scores[write_item_idx] =
+            kept_bboxinfo_flag[cur_idx]
+                ? T_SCORE(beforeNMS_scores[read_item_idx])
+                : T_SCORE(score_shift);
+        afterNMS_index_array[write_item_idx] =
+            kept_bboxinfo_flag[cur_idx] ? loc_bboxIndex[t] : -1;
+      }
+    }
+  }
+}
```
Comment (Xreki) [inline paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu:539]: 这个函数有148行，影响阅读，请考虑下进一步封装

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu [hunk 8]
```diff
@@ -0,0 +1,1141 @@
...
+                  void* workspace,
+                  bool isNormalized,
+                  bool confSigmoid,
+                  bool clipBoxes,
+                  int scoreBits,
+                  bool caffeSemantics) {
+  PADDLE_ENFORCE_EQ(
+      shareLocation,
+      true,
+      paddle::platform::errors::Fatal("shareLocation=false is not supported."));
```
Comment (Xreki) [inline paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu]: 使用`phi::errors`代替，错误类型用`Unsupported`

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu [hunk 9]
```diff
@@ -0,0 +1,1141 @@
...
+  size_t bboxDataSize = detectionForwardBBoxDataSize<T>(N, perBatchBoxesSize);
+  void* bboxDataRaw = workspace;
+  PADDLE_ENFORCE_GPU_SUCCESS(cudaMemcpyAsync(
+      bboxDataRaw, locData, bboxDataSize, cudaMemcpyDeviceToDevice, stream));
+  void* bboxData = bboxDataRaw;
+
+  const int numScores = N * perBatchScoresSize;
```
Comment (Xreki) [inline paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu]: 变量名请采用`num_scores`形式

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu [hunk 10]
```diff
@@ -0,0 +1,1141 @@
...
+  bool return_index = index != nullptr;
+  bool has_roisnum = rois_num.get_ptr() != nullptr;
+  auto score_dims = scores.dims();
+  auto score_size = score_dims.size();
+
+  bool is_supported = (score_size == 3) && (nms_top_k >= 0) &&
+                      (nms_top_k <= 4096) && (keep_top_k >= 0) &&
+                      (nms_eta == 1.0) && !has_roisnum;
+  if (!is_supported) {
+    VLOG(3)
```
Comment (Xreki) [inline paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu]: log级别设置到6以上

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu [hunk 11]
```diff
@@ -0,0 +1,1141 @@
...
+        << ", has_roisnum=" << has_roisnum;
+
+    DenseTensor bboxes_cpu, scores_cpu, rois_num_cpu_tenor;
+    DenseTensor out_cpu, index_cpu, nms_rois_num_cpu;
+    paddle::optional<DenseTensor> rois_num_cpu(paddle::none);
+    auto cpu_place = phi::CPUPlace();
+    const auto gpu_place = ctx.GetPlace();
+
+    // copy from GPU to CPU
+    paddle::framework::TensorCopy(bboxes, cpu_place, &bboxes_cpu);
```
Comment (Xreki) [inline paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu]: 用phi下面的函数代替

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu [hunk 12]
```diff
@@ -0,0 +1,1141 @@
...
+  NonZeroKernel<int, Context>(ctx, nmsed_valid_mask, &valid_indices);
+
+  const int64_t valid_samples = valid_indices.dims()[0];
+  out->Resize({valid_samples, 6});
+  ctx.template Alloc<T>(out);
+  phi::funcs::GPUGatherNd<T, int64_t>(ctx, raw_out, valid_indices, out);
+  index->Resize({valid_samples, 1});
+  ctx.template Alloc<int>(index);
+  phi::funcs::GPUGatherNd<int, int64_t>(
+      ctx, nmsed_indices, valid_indices, index);
```
Comment (Xreki) [inline paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu:1134]: 这个函数有197行代码，影响阅读，请考虑下进一步封装

File: python/paddle/fluid/tests/unittests/test_multiclass_nms_op.py (hunk 2 above)
Comment (Tom-Zheng) [reply python/paddle/fluid/tests/unittests/test_multiclass_nms_op.py:807]: 理由同下。

File: python/paddle/fluid/tests/unittests/test_multiclass_nms_op.py (hunk 3 above)
Comment (Tom-Zheng) [reply python/paddle/fluid/tests/unittests/test_multiclass_nms_op.py]: GPU kernel和CPU 的输出bounding box的顺序不同： CPU是先按score降序，再按class升序；GPU输出是按score降序。所以需要先sort之后才好做比较。

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu (hunk 5 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu]: Done.

File: paddle/fluid/operators/detection/multiclass_nms_op.cc (hunk 4 above)
Comment (Tom-Zheng) [reply paddle/fluid/operators/detection/multiclass_nms_op.cc:619]: 这里是因为https://github.com/PaddlePaddle/Paddle/blob/82779c79c97883cd3d31fee7b6311f6d89afebde/paddle/fluid/operators/detection/multiclass_nms_op.cc#L119 写死了返回CPU kernel， 所以这里overload让它支持GPU kernel. 跟https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/phi/api/yaml/legacy_ops.yaml#L129 似乎无关？

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu (hunk 11 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu]: Done

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu (hunk 6 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu]: Done

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu (hunk 8 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu]: Done

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu (hunk 9 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu]: Done

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu (hunk 10 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu]: Done

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu (hunk 7 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu:539]: Done

File: paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu (hunk 12 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/gpu/multiclass_nms3_kernel.cu:1134]: 其实大多数都是在做输入的准备以及输出的后处理，参数比较多所以显得长，我觉得不太好再封装了。我加了一些注释，请看是否可以。

File: paddle/fluid/operators/detection/multiclass_nms_op.cc (hunk 4 above)
Comment (Xreki) [reply paddle/fluid/operators/detection/multiclass_nms_op.cc:619]: phi目录下的kernel是multiclass_nms3，这里重写multiclass_nms3的`GetExpectedKernelType`，也是为了指定依据哪个输入的数据类型来选Kernel。

File: python/paddle/fluid/tests/unittests/test_multiclass_nms_op.py (hunk 3 above)
Comment (Xreki) [reply python/paddle/fluid/tests/unittests/test_multiclass_nms_op.py]: > GPU kernel和CPU 的输出bounding box的顺序不同： CPU是先按score降序，再按class升序；GPU输出是按score降序。所以需要先sort之后才好做比较。

需要请算子负责人、或者更熟悉这个算子的人来review下这部分逻辑。

就单测本身而言，OpTest框架会同时检查静态图、动态图，新增的这个单测只检查了静态图。仍然建议基于原单测框架实现单测。

Comment (onecatcn) [reply python/paddle/fluid/tests/unittests/test_multiclass_nms_op.py]: @lyuwenyu 请review部分的逻辑

Comment (lyuwenyu) [reply python/paddle/fluid/tests/unittests/test_multiclass_nms_op.py]: gpu和cpu处理逻辑是有diff

Comment (onecatcn) [reply python/paddle/fluid/tests/unittests/test_multiclass_nms_op.py]: > 

那PR中这个方案是OK的吗？如果不OK，请问有什么修改意见吗？ @lyuwenyu

Comment (lyuwenyu) [reply python/paddle/fluid/tests/unittests/test_multiclass_nms_op.py]: 我觉得没问题的

Comment (Tom-Zheng) [reply python/paddle/fluid/tests/unittests/test_multiclass_nms_op.py]: Done. 已修改使用单测框架

**Review by lyuwenyu — APPROVED** [review]

**Review by lyuwenyu — APPROVED** [review]
lgtm

**Review by Xreki — APPROVED** [review]
LGTM

**Review by qili93 — APPROVED** [review]
LGTM for @unittest.skipIf

**Review by jerrywgz — APPROVED** [review]
LGTM

**Comment by Tom-Zheng** [comment]
@XiaoguangHu01 Would you please review this PR?

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Comment by Tom-Zheng** [comment]
@Xreki  I think we are OK to merge here.

**Pull request merged**